hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/scm/client/OMBlockPrefetchClient.java

```diff
@@ -0,0 +1,296 @@
+    int allowedPrefetch = Math.max(0, MAX_PARALLEL_CACHE_BUILDUP - parallelBlocks);
+    int finalBlocksToFetch = Math.min(Math.min(numBlocks, availableSpace), allowedPrefetch);
+
+    if (finalBlocksToFetch > 0) {
+      PARALLEL_ADDITIONAL_BLOCKS.addAndGet(finalBlocksToFetch);
+    }
+
+    return finalBlocksToFetch;
+  }
+
+  @SuppressWarnings("parameternumber")
+  public List<AllocatedBlock> getBlocks(long scmBlockSize, int numBlocks, ReplicationConfig replicationConfig,
+                                        String serviceID, ExcludeList excludeList, String clientMachine,
+                                        NetworkTopology clusterMap) throws IOException {
+    long readStartTime = Time.monotonicNowNanos();
+    List<AllocatedBlock> allocatedBlocks = new ArrayList<>();
+    int retrievedBlocksCount = 0;
+    ConcurrentLinkedDeque<ExpiringAllocatedBlock> queue = blockQueueMap.get(replicationConfig);
+
+    // We redirect to the allocateBlock RPC call to SCM when we encounter an untested ReplicationConfig or a populated
+    // ExcludeList, otherwise we return blocks from cache.
+    if (queue != null && excludeList.isEmpty()) {
+      while (retrievedBlocksCount < numBlocks) {
+        ExpiringAllocatedBlock expiringBlock = queue.poll();
+        if (expiringBlock == null) {
+          break;
+        }
+
+        if (System.currentTimeMillis() > expiringBlock.getExpiryTime()) {
+          continue;
+        }
+
+        AllocatedBlock block = expiringBlock.getBlock();
+        List<DatanodeDetails> sortedNodes = sortDatanodes(block.getPipeline().getNodes(), clientMachine, clusterMap);
+        if (!Objects.equals(sortedNodes, block.getPipeline().getNodesInOrder())) {
+          block = block.toBuilder()
+              .setPipeline(block.getPipeline().copyWithNodesInOrder(sortedNodes))
+              .build();
+        }
+        allocatedBlocks.add(block);
+        retrievedBlocksCount++;
+      }
+
+      int remainingBlocks = numBlocks - retrievedBlocksCount;
+
+      // If there aren't enough blocks in the cache, we make an RPC call to SCM and prefetch additional batch max + the
+      // requested blocks in the same call, avoiding a separate prefetchBlocks call to SCM.
+      if (remainingBlocks > 0) {
+        int overAllocateBlocks = blocksToFetch(queue.size(), ADDITIONAL_BLOCKS_MAX);
+        List<AllocatedBlock> newBlocks =
+            scmBlockLocationProtocol.allocateBlock(scmBlockSize, remainingBlocks + overAllocateBlocks,
+                replicationConfig, serviceID, excludeList, clientMachine);
+        long expiryTime = System.currentTimeMillis() + expiryDuration;
+        int newBlocksSize = newBlocks.size();
+        allocatedBlocks.addAll(newBlocks.subList(0, remainingBlocks));
+
+        if (overAllocateBlocks > 0) {
+          for (AllocatedBlock block : newBlocks.subList(remainingBlocks, newBlocksSize)) {
+            queue.offer(new ExpiringAllocatedBlock(block, expiryTime));
```

Comment:
Make sure there are good boundary tests and tests to check there no overlap between blocks returned to the client and blocks added to the cache.